File: 3349-maximum-length-substring-with-two-occurrences/maximum-length-substring-with-two-occurrences.cpp
class Solution {
public:
    int maximumLengthSubstring(string s) {
        int l=0,r=0,maxlen=0,countExceed=0;
        unordered_map<char, int> charCountMap;
        while(r<2){
            auto it = charCountMap.find(s[r]);
            if(it != charCountMap.end()){
                it->second++;
            }else{
                charCountMap[s[r]] =1;
            }
            maxlen++;
            r++;
        }
        while(r<s.size()){
            auto it = charCountMap.find(s[r]);
            if(it != charCountMap.end()){
                it->second++;
                if(it->second==3){
                    countExceed++;
                }
            }else{
                charCountMap[s[r]] =1;
            }
            if(countExceed>0){
                it = charCountMap.find(s[l]);
                it->second--;
                if(it->second ==2){
                    countExceed --;
                }
                l++;
            }else{
                maxlen = max(maxlen,r-l+1);
            }
            r++;
        }
        return maxlen;
    }
};

File: 3349-maximum-length-substring-with-two-occurrences/maximum-length-substring-with-two-occurrences.go
func maximumLengthSubstring(s string) int {
    var l,r,maxLen,exceedCount int =0, 0, 0, 0

    charCountMap:= make(map[string]int)

    for r<2 {
        value, exists := charCountMap[string(s[r])]
        if exists {
            value++
            charCountMap[string(s[r])] = value
        }else{
           charCountMap[string(s[r])] = 1
        }
        maxLen++
        r++
    }
    for r< len(s){
        value, exists := charCountMap[string(s[r])]
        if exists {
            value++
            charCountMap[string(s[r])] = value
            if charCountMap[string(s[r])] ==3{
                exceedCount++
            }
        }else{
           charCountMap[string(s[r])] = 1
        }

        if(exceedCount>0){
            charCountMap[string(s[l])]--
            if charCountMap[string(s[l])] ==2 {
                exceedCount--
            }
            l++
        }else{
            maxLen = maxInt(maxLen,r-l+1)
        }
        r++
        
    }
    return maxLen


}

func maxInt(a, b int) int {
    if a > b {
        return a
    }
    return b
}